pkg/identity/cidr.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/labels/cidr"
+)
+
+// AllocateCIDRIdentities allocates identities for each of the specified CIDR
+// prefixes so they will be available later during policy resolution.
+//
+// On success, the returned slice will have a 1-to-1 correspondence with the
+// CIDR in the provided prefixes slice, eg the identity for prefixes[0] will
+// be held in identities[0].
+// On failure, attempts to clean up after itself and returns the error.
+func AllocateCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	if len(prefixes) == 0 {
+		return nil, nil
+	}
+
+	log.Debugf("Attempting to allocate identities for %d prefixes", len(prefixes))
+
+	// Allocate identities for each CIDR.
+	res = make([]*Identity, 0, len(prefixes))
+	for _, prefix := range prefixes {
+		if prefix == nil {
+			continue
+		}
+		labels := cidr.GetCIDRLabels(prefix)
+		id, _, err := AllocateIdentity(labels)
```

Comment:
Don't we need to track `isNew` here? (the _ return value). Overlapping CIDRs can generate the same identity (one with a more specific range, I think) but then, when we clean up on error below, we would try to release it even if the allocate was a no-op in the KVStore (thus releasing that ID globally).

Reply:
Oh, the kvstore code is refcounting for us. This is probably fine.

Reply:
Oh, and overlapping CIDRs have different sets of labels, so they have different identities.